planner/core/rule_partition_processor.go

```diff
@@ -1575,13 +1576,32 @@ func (p *rangeColumnsPruner) partitionRangeForExpr(sctx sessionctx.Context, expr
 		return 0, len(p.data), false
 	}
 
+	exprCollation, err := expression.CheckAndDeriveCollationFromExprs(sctx, opName, col.RetType.EvalType(), con, p.data[0])
+	if err != nil {
+		return 0, len(p.data), false
+	}
+	if exprCollation.Coer < expression.CoercibilityImplicit {
+		// Explicit collation, if set to Binary collation, we can still use prune partitions,
+		// but we need to change the constants collation
+		conCharSet, conColl := con.CharsetAndCollation()
+		colCharSet, colColl := p.partCol.RetType.Charset, p.partCol.RetType.Collate
+		if !collate.IsBinCollation(conColl) || conCharSet != colCharSet {
+			return 0, len(p.data), true
+		}
+		clonedCon := con.Clone()
+		if newCon, ok := clonedCon.(*expression.Constant); ok {
+			con = newCon
+		}
+		con.SetCharsetAndCollation(conCharSet, colColl)
+	}
```

Comment:
I think we can fall back to all partitions if the collation used in the expression is not the same as the partition column collation.

And for the `_bin` collation, we can only consider the `eq` expression. MySQL does not prune anything if not `eq`
```sql
mysql> create table t (a varchar(255) charset utf8mb4 collate utf8mb4_general_ci) partition by range columns (a) (partition puppera values less than ("AAA"), partition plowera values less than ("aaaa"), partition pp values less than ('hhhh'));
Query OK, 0 rows affected (0.04 sec)

mysql> insert into t values ("aa");
Query OK, 1 row affected (0.02 sec)

mysql> explain select * from t where a > "AAAa" collate utf8mb4_bin;
+----+-------------+-------+--------------------+------+---------------+------+---------+------+------+----------+-------------+
| id | select_type | table | partitions         | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra       |
+----+-------------+-------+--------------------+------+---------------+------+---------+------+------+----------+-------------+
|  1 | SIMPLE      | t     | puppera,plowera,pp | ALL  | NULL          | NULL | NULL    | NULL |    1 |   100.00 | Using where |
+----+-------------+-------+--------------------+------+---------------+------+---------+------+------+----------+-------------+
1 row in set, 1 warning (0.00 sec)

mysql> explain select * from t where a = "AAAa" collate utf8mb4_bin;
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra       |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
|  1 | SIMPLE      | t     | pp         | ALL  | NULL          | NULL | NULL    | NULL |    1 |   100.00 | Using where |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
1 row in set, 1 warning (0.00 sec)
```

Reply:
Thank you for finding this issue! I added tests and the restriction to only support `eq` if explicit _bin collation which is different from partition column/definition.